File: 001/gradient_descent_minimization.jl
function f(x, y)
    return 1 + 2 * (x * x) + 3 * (y * y)
end

delta = 0.1
alpha = 0.001

function run(x, y)
    fx = (f(x + 0.5 * delta, y) - f(x - 0.5 * delta, y)) / delta
    fy = (f(x, y + 0.5 * delta) - f(x, y - 0.5 * delta)) / delta
    x -= alpha * fx
    y -= alpha * fy

    return (x, y)

end

start_time = time_ns()

for n in 1:10
    update_1 = run(1,2)
    update_2 = run(update_1[1], update_1[2])

end

final_time = time_ns() - start_time

print("Julia: ", final_time/10000000000)

File: 001/gradient_descent_minimization.py
import time

delta = 0.1
alpha = 0.001

def f(x, y):
    return 1 + 2 * (x ** 2) + 3 * (y ** 2)

def run(x, y):
    fx = (f(1 + 0.5 * delta, 2) - f(1 - 0.5 * delta, 2)) / delta
    fy = (f(1, 2 + 0.5 * delta) - f(1, 2 - 0.5 * delta)) / delta
    x -= alpha * fx
    y -= alpha * fy

    return (x, y)

start_time = time.time()

for num in range(10):
    update_1 = run(1, 2)
    update_2 = run(update_1[0], update_1[1])

print("Python:", (time.time() - start_time) / 10)